x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/ingest/SetSecurityUserProcessor.java

```diff
@@ -131,6 +135,14 @@ public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
                     if (apiKeyId != null) {
                         apiKeyField.put("id", apiKeyId);
                     }
+                    final Object apiKeyMetadata = authentication.getMetadata().get(ApiKeyService.API_KEY_METADATA_KEY);
+                    if (apiKeyMetadata != null) {
+                        final Tuple<XContentType, Map<String, Object>> tuple =
+                            XContentHelper.convertToMap((BytesReference) apiKeyMetadata, false, XContentType.JSON);
+                        if (false == tuple.v2().isEmpty()) {
+                            apiKeyField.put("metadata", tuple.v2());
+                        }
```

Comment:
I'd like to see all of this moved to the ApiKeyService rather than being duplicate everywhere we need to use API key metadata.

```suggestion
                    final Map<String,Object> apiKeyMetadata = ApiKeyService.getApiKeyMetadata(authentication);
                    if (apiKeyMetadata.isEmpty() == false) {
                        apiKeyField.put("metadata", apiKeyMetadata);
                    }
```